[Jung] 미션3 : 체스판 초기화 PR입니다

rla36 opened the pull request:
안녕하세요 Jung입니다!
미션3 PR 작성 시점에 미션2 PR이 머지 되지 않아서 커밋이 많아보입니다.
**Pawn의 Representation추가 …(03a5e8a)부터 미션 3 커밋이니** 리뷰하실 분들은 여기서부터 봐주시면 감사하겠습니다.

- 이번 미션에서는 호눅스가 얘기하신대로 커밋 메세지의 초점을 '무엇을'이 아닌 '왜'에 맞추도록 노력했습니다.
또한 커밋 컨벤션에 맞추어 feat/test 등 형식을 통일하려했는데 이 부분은 잘 되지 않았네요..ㅎㅎ

**Pawn**
- 저는 pawn의 생성자에 representation없이 색깔만 받도록했습니다.  Representation이 pawn의 색상에 따라 흰색이면 소문자 p 검은색이면 대문자 P인 것이 확실한 상황이기 때문에, Pawn(Color.WHITE, 'p')나 Pawn(Color.BLACK, 'P')와 같이 색깔과 representation을 둘다 받는 것이 중복이고, 비효율적이라는 생각이 들었기때문입니다.
white/black 를 enum으로 다루기 때문에 생성자에는 이 두 값 외에는 들어올 수 없기에 Pawn의 생성자로 색깔만 받아 입력된 색에 따라 representation을  갖도록 했습니다. 이렇게 하니 생성자의 representation 부분에 정해진 알파벳 외의 알파벳이 들어가거나, 상충하는 Color와 Representation을 입력했을 경우를 고려하지 않아도 되는 장점이 있습니다!

**Board**
- pawn을 관리하는 ArrayList인 pawnList을 whitePawnList / blackPawnList로 쪼개었고, 그에 맞춰 필요한 경우 메서드들도 각 색깔의 폰 리스트를 관리하도록 나누었습니다.
- 미션에는 없는 reset()메서드를 구현했습니다. initialize를 하면 white/black PawnList에 8개씩 add해서 초기화를 합니다.  그런데 add메서드에 pawn개수를 제한하는 예외처리를 할 경우, Main에서 사용자가 start을 여러 번 누르는 경우에 initialize가 여러번 되어 pawnList가 각각 8개를 초과해 예외가 발생합니다. 예상하기에 start를 여러번 누르는 것이 있을수 있는 일이고(?) 그때마다 게임을 예외와 함께 종료시키고 싶지 않아 start를 누르면 그때마다 white/black PawnList를 clear()하는 방식으로 pawnList의 폰이 8개를 넘는 일이 없도록 했습니다. 

그 외에는 미션 요구사항에 명시된 대로 진행했습니다.
감사합니다!

rla36 committed:
기본생성자로 흰색폰생성 검증하는 테스트 메서드 작성

rla36 committed:
흰색폰 생성하는 기본생성자 추가

rla36 committed:
Board에 폰리스트와 add메서드 추가

Board클래스 생성 후 Board에서 폰 목록을 관리할 수 있는 ArrayList를
pawnList라는 이름으로 만들었습니다.
add(Pawn pawn)을 통해 보드의 pawnList에서 폰들을 관리할 수 있게
했습니다.

rla36 committed:
Board의 pawn개수 리턴하는 size() 메서드 추가

rla36 committed:
Board의 index번째 pawn을 리턴하는 findPawn()메서드 추가

rla36 committed:
Board의 메서드 검증하는 Test클래스 추가

Board의 add(Pawn) size() getPawn(index)를 검증하는 테스트 메서드들
작성했습니다.

rla36 committed:
패키지 분리

Pawn과 PawnTest는 pieces 패키지에 포함시켰습니다.
나머지는 chess 패키지에 두었습니다.

rla36 committed:
size findPawn메서드 테스트 중복제거

rla36 committed:
Pawn의 Representation추가

Pawn의 인스턴스 변수로 char 타입 representation을 추가했습니다.
미션과 다르게 Pawn의 생성자 매개변수로 Color만 받도록했습니다.  그 이유는
흰색폰이면 p, 검은색 폰이면 P의 representation을 가질 것이 이미
확실히 정해져있는 상황에서 생성자 매개변수로 색깔에 더해  p/P를 추가로  받는 것은 불필요하다고 생각했기 때문입니다.

Color Enum을 사용함으로써 생성자에 WHITE BLACK 두가지 색만을 쓰도록 강제했기 때문에 Pawn의 representation도 enum Color에 묶음으로써 p/P만 쓰이도록 강제할 수 있는 효과를 기대했습니다.
매개변수에 p/P 외의 문자가 들어오지 않도록 보장하고, 16개나 되는 폰을
생성할 때 피로하지 않도록 pawn 생성자의 매개변수를 Color 하나만 받도록
했으니 참고부탁드립니다..

rla36 committed:
Pawn의 getRepresentation 메서드 검증하는 테스트 추가

루카스와 달리 Pawn의 생성자 매개변수로 Color만 받도록 했기 때문에
테스트메서드 역시 다르게 작성했습니다.

rla36 committed:
test로 디렉토리 이름 수정

rla36 committed:
BoardTest내 멤버변수 접근제어자 private으로 수정

BoardTest 내에서만 쓰이는 변수들이기 때문에 접근제어자를 private으로
수정했습니다.

rla36 committed:
assertAll 하나만 쓰이도록 test  람다식으로 수정

rla36 committed:
기본생성자 위치 수정

기본생성자가 매개변수가 있는 생성자의 위로 오도록 위치 수정했습니다.

rla36 committed:
패키지 정리

net.jung.chess 디렉토리  아래 pieces 디렉토리를 비롯한 클래스/enum들이 위치하도록
재정리했습니다.

rla36 committed:
PawnList를 whitePawnList와 blackPawnList로 분리

PawnList를 흰색/검은색별로 관리하기 위해 분리했습니다.
보드에서 흰색과 검은색 폰을 한꺼번에 다뤄야할 일이 없을 것으로 예상되어
이렇게 했고, findPawn메서드와 테스트메서드들도 이에 맞춰 수정했습니다.

rla36 committed:
initialize메서드 추가해 보드에 흰/검 폰 8개씩 추가

initialize()메서드로 보드의 whitePawnList와 blackPawnList에 각각 흰색,
검은색 폰을 8개씩 올리도록 했습니다.
getWhitePawnsResult(), getBlackPawnsResult로는 흰/검폰 ArrayList의 폰 개수만큼
폰을 대표하는 알파벳을 나열해 String으로 리턴하도록 했습니다.

rla36 committed:
test: initialize메서드 검증 테스트 추가

initialize()메서드로 보드에 흰/검 폰을 8개씩 추가한 후
getWhitePawnsResult(), getBlackPawnsResult()로 폰들의 대표 알파벳을 합쳐
얻은 값이 기댓값과 일치하는지 검증하는 테스트를 작성했습니다.

rla36 committed:
getRepresentation의 중복 코드 제거

getWhiteRepresentation, getBlackRepresentation의 중복 코드를 하나의
메서드로 빼내어 중복 코드를 제거했습니다.

rla36 committed:
feat: print() 추가

체스판의 상황을 보여주는 string을 반환하는 print()메서드를 구현했습니다.
늘어나는 체스 말을  관리하기 위해 xy 좌표를 추가하면 좋을 것
같았으나, 현재 진행단계에서는 비어있는 곳과 말들의 위치를 string 안에
하드코딩해서 표현했습니다.

rla36 committed:
test: print()메서드 검증 테스트 추가

print()로 리턴되는 스트링이 기댓값과 같은지 검증하는 테스트를
추가했습니다. 터미널에 출력되는 것을 눈으로 확인하고, 스트링 기댓값과
일치하는지 assertThat을 통해 확인하도록 했습니다.

rla36 committed:
feat: Board에 reset메서드 추가해 초기상태로 되돌리도록함

Main에서 start/end 입력값을 받을때 start를 연속적으로 받을 경우
initialize가 여러번 수행되어 whitePawnList와 blackPawnList에 흰/검 폰이
8개씩 계속 추가되는 문제가 발생했습니다. initialize가 호출하는  add메서드에 8개이상 추가되지 못하도록 예외처리를 하는 방법도 고려했으나, 게임 유저가 start 여러번 누르는 일도 자주 있을 수 있는 일이라는 생각이 들어 게임을 종료시키지 않는 방법을 찾았습니다. start을 누를때마다 기존 흰/검 pawnList을 초기상태로 되돌린 후에 initialize를 하면 start여러번 눌러도 white/blackPawnList가 각각 8개를 넘어서는 일이 일어나지 않기 때문에 reset메서드를 추가했습니다.

rla36 committed:
사용자 입력 받아 게임 시작/종료하는 기능 추가

Scanner으로 사용자의 입력을 받아 게임을 시작/종료하도록 했습니다.
start입력시 게임 시작, end 입력시 게임 종료하도록 했고, 공백을 비롯해
잘못된 명령어를 입력하면 다시 입력하도록 했습니다.

SANGCO commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -0,0 +1,76 @@
+public class Board {
+    private List<Pawn> blackPawnList = new ArrayList();
+    private List<Pawn> whitePawnList = new ArrayList();
+    private final int MAX_PAWNS_SIZE = 8;
+    private String blankRow = "********";
```
클래스의 변수는 최소화 하려고 노력해야 한답니다.
이 친구들은 인스턴스 변수일 필요가 있을까요?
지역 변수로 선언하는게 좋지 않을지 한번 고민해봅시다!

MAX_PAWNS_SIZE의 경우도 다른 클래스에서도 사용을 한다면 
static을 붙여서 상수로 관리하면서 여기저기서 쓰는건 좋은거 같습니다.
근데 현재는 이 클래스에서만 사용하는거 같은데 맞나요?

rla36 replied:
넵 말씀해주신대로 수정했습니다! 뭔가 미래에 같은 클래스 내에서 또 쓰일 것만 같아서ㅋㅋㅋ 일단 인스턴스 변수로 선언해뒀었는데 나중에 다른 곳에 쓰이게 되면 그때 인스턴스 변수로 바꾸더라도 처음 만들때에는 쓰이는 곳에서만 쓰이도록 메서드 안에 지역변수로 넣는 것이 좋군요!! 너무 나중을 생각해도 안되겠어요 말씀 감사합니다!

SANGCO commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -0,0 +1,76 @@
+    private final int MAX_PAWNS_SIZE = 8;
+    private String blankRow = "********";
+
+    public void add(Pawn pawn) {
+        if (pawn.getColor() == (Color.WHITE)) {
+            whitePawnList.add(pawn);
+            return;
+        }
+        blackPawnList.add(pawn);
+    }
```
저는 개인적으로 리턴 타입을 void하고 의미 없는 return으로 
메서드를 종료하는걸 별로 안좋아 하는데요.
addWhitePawn(), addBlackPawn() 이런식으로 메서드 분리를 시도해 봅시다!

추가로 addWhitePawn()을 호출하면서 파라미터로 블랙 Pawn을 넘기는 테스트 코드도 만들어 봅시다~

rla36 replied:
pawn만 더하면 알아서 컬러를 판단해서 white/black pawnList에 더해주는 메서드를 구현하고 싶어서 저렇게 했습니다! 그런데 pawn을 add 할때 어떤 색인지 확실히 아는 경우가 많을 것 같아 addWhitePawn addBlackPawn으로 한기능만 담당하도록 메서드를 분리하는게 맞을 것 같네요.. 수정했습니다
그리고 addWhitePawn(Pawn blackPawn) 과 addBlackPawn(Pawn whitePawn)했을 떄 
각 whitePawnList와 blackPawnList에 더해지지 않는 것을 확인하는 테스트 메서드도 작성했습니다.

혹시 void 메서드를 return으로 종료하는 것 선호하지 않으시는 이유 여쭤도 되나용? else를 쓰지 않기 위해 저렇게 한 것이라 궁금합니다!

SANGCO replied:
void 메서드에 return 안쓰는건 지극히 제 개인 취향입니다ㅋ
저는 리턴 타임이 void인데 return이 있는게 보기 싫더라구요.
위와 같은 상황에 저는 void return 하느니 if문을 하나 더 넣던지 메소드 분리를 할거 같네요.
호눅스는 어떠세요?ㅋ

honux77 replied:
객체지향체조 원칙에서 else를 쓰지 않으려고 해서 저럴 때는 return을 쓰게 되는 것 같아요.
return은 원래 값을 전달한다와 함수를 끝낸다라는 두 가지 용도가 있기 때문에 저는 적절하다고 생각합니다. 
상코는 아마 첫번째 의미만 생각하고 있는 것 같아요. 상코 말도 일리가 있는게 return이 너무 많으면 또 가독성이 떨어지거든요. 

지금 상황에서는 return 하나 정도니까 무난한 것 같습니다.

SANGCO commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -0,0 +1,76 @@
+
+    public String getPawnsResult(List<Pawn> pawnList) {
+        StringBuilder pawnReps = new StringBuilder();
+        int colorPawnSize = pawnList.size();
```
저는 이런식으로 지역 변수로 빼는 부분은 좋은거 같아요.
그런데  이 경우에는 list에 size()를 가지고 오는것도 코드가 잘 읽히는데
굳이 지역 변수로 뺄 필요가 있을까 하는 생각이 드네요.
이거는 개인적인 의견입니다. 수정 안하셔도 됩니다ㅋ

rla36 replied:
for문 안에서 i가 +1 될때마다 pawnList.size() 연산을 하는 것이 싫어서 이렇게 했습니다..ㅎㅎ 
복잡한 연산일 경우엔 for문 바깥에 빼는 것이 좋지만 size()같은 간단한(?) 연산의 경우에는 그냥 for문 안에 넣어도 성능에 큰 차이가 없나보군요!

SANGCO replied:
> for문 안에서 i가 +1 될때마다 pawnList.size() 연산을 하는 것이 싫어서 이렇게 했습니다..ㅎㅎ

훌륭하십니다. 저는 보통 그냥 써서 생각을 못해봤네요ㅋ
[Performance Comparison – Different For Loops in Java](https://howtodoinjava.com/java/collections/performance-comparison-of-different-for-loops-in-java/)
매번 연산을 하는건 아니고 ArrayList 클래스 안에 있는 size 필드를 가지고 온답니다. (ArrayList 한번 들어가서 봅시다.)

이펙티브 자바 `아이템 57 지역변수의 범위를 최소화하라` 마지막 부분에 나오는 내용인데

```java
for (int i = 0, n = expensiveComputation(); i < n; i++) {
    ... // i로 무언가를 한다.
}
```

진짜 성능에 문제가 있다면 저런식으로 하는것도 좋겠죠?

rla36 replied:
for문의 (초기화; 조건문; 증감)에서 초기화 부분에 연산을 넣으면 하면 for문이 돌때마다 연산하지 않는군요!!
저렇게 쓰면 간결성과 성능을 동시에 잡을수있네요.. 앞으로 이렇게 해야겠어요 
effective java는 아직 구입하지 못했는데 빨리 공부 시작해야겠습니다 감사해요!

SANGCO commented on src/test/java/net/jung/chess/BoardTest.java:
```diff
@@ -0,0 +1,61 @@
+    void checkPrint(){
+        board.initialize();
+        System.out.print(board.print());
+        String expectedLayout = "********\nPPPPPPPP\n********\n********\n********\n********\npppppppp\n********";
```
여기도 메서드로 빼서 StringBuilder로 만들어 놓으면 향후에 편할거 같지 않을까요?

rla36 replied:
조금 더 자세히 말씀해주시면 감사하겠습니당 ㅜ 
expectedLayout 스트링을 메서드 안에 선언한 이유는 white/black 폰들만 놓인 상태의 보드 기댓값 String 이 
이 테스트 메서드 안에서만 쓰일 것으로 예상했기 때문입니다! 앞으로는 다른 말들도 보드 위에 놓일 것이고 보드 상태를 확인하는 테스트메서드를 추가하더라도 흰/검 폰만 놓인 상태의 스트링은 여기서밖에 안쓰일것같았습니다.
initialize()로 호출되는 getWhitePawnsResult() getBlackPawnsResult가 저 expectedLayout과 일치하는지 확인하는 것이 테스트의 목적이기 때문에 expectedLayout 스트링을 StringBuilder 메서드로 빼내어도 getWhitePawnsResult, getBlackPawnsResult 메서드를 사용할 수 없습니다. 그렇게 되면 getWhitePawnsResult, getBlackPawnsResult을 자기자신과 비교하게 되어 항상 test passed일것이기 때문입니다. 
제가 상코님 말씀을 제대로 이해한것인지 확실하지 않은데,  제가 코드를 이렇게 작성한 이유는 이러하고 어떻게 고치면 좋을지 추가적인 설명 해주시면 감사하겠습니다ㅎㅎ

SANGCO replied:
Jung님 말씀이 맞는거 같아요ㅋ
저는 StringBuilder로 만들어 놓으면 더 수정하기 쉬울거라고 생각했는데
다시 생각해보니깐 뭐 그게 그건거 같네요.
이 부분은 이대로 가도 될거 같습니다~

SANGCO commented on src/main/java/net/jung/chess/Main.java:
```diff
@@ -0,0 +1,34 @@
+
+public class Main {
+    public static void main(String[] args) {
+        Scanner sc = new Scanner(System.in);
```
Scanner와 같이 AutoCloseable을 구현한 클래스의 경우 
[try-with-resources](https://www.baeldung.com/java-try-with-resources)를 사용해서 쉽게 리소스를 해제할 수 있답니다.
이번 미션에서 try-with-resources 관련되어서 학습하고 적용해보면 좋을거 같아요~

rla36 replied:
자바의 정석에서 봤었는데 생각도 못했어요 ㅋㅋ 한번 시도해봤습니다 감사해요!

SANGCO commented on src/main/java/net/jung/chess/pieces/Pawn.java:
```diff
@@ -0,0 +1,26 @@
+public class Pawn {
+    private Color color;
+    private char representation;
+
+
+
```
여기는 왜 세칸이나 공백이ㅋ

SANGCO commented on src/main/java/net/jung/chess/pieces/Pawn.java:
```diff
@@ -0,0 +1,26 @@
+
+
+
+    public Pawn(){
```
Pawn 클래스는 전체적으로 코드가 정렬이 안되어 있네요ㅋ
Pawn() 한칸 띄우고 {
IntelliJ에 Reformat code 옵션을 찾아서 설정 해보는건 어떨까요.

rla36 replied:
넵 Ctrl+Alt+l 하니 자동으로 예쁘게 정렬되네요 보는 사람이 편하게 읽도록 최대한 예쁘게 정렬해야겠습니다!! 아래 코멘트 주신 것과 함께 수정했습니다~

SANGCO commented on src/main/java/net/jung/chess/pieces/Pawn.java:
```diff
@@ -0,0 +1,26 @@
+    }
+    public Pawn(Color color) {
+        this.color = color;
+        representation = color.equals(Color.WHITE)? 'p':'P';
```
여기도 코드 정렬 `color.equals(Color.WHITE) ? 'p' : 'P'`

'p', 'P' 이렇게 처리하시는건 좋은거 같아요.
미션3의 요구사항을 구현하는데 이정도면 충분한거 같습니다. 굿굿

rla36 replied:
수정했씀다!

honux77 replied:
좋네요!

SANGCO reviewed (changes requested):
Jung님 구현 잘하셨네요. 고생하셨습니다!
리뷰 남겼으니 확인해 보시고 질문 있으시면 댓글 부탁드립니다~

rla36 committed:
Pawn클래스 코드 정렬

rla36 committed:
인스턴스변수 메서드 안의 지역변수로 편입

특정 메서드에서만 쓰이는 멤버변수를 메서드 안으로 포함시켰다.

rla36 committed:
add메서드 addWhitePawn과 addBlackPawn으로 분리

기존 add메서드는 매개변수로 입력받는 pawn의 색깔을 메서드 안에서
판단해서 white/blackPawnList에 추가했습니다. 메서드별로 기능  하나씩을
담당하는 것이 좋고, pawn을 추가할 때 그 색을 알고 있는 경우가 많을 것
같아 색깔별로 메서드를 분리했습니다.

rla36 committed:
test: addWhite/BlackPawn 매개변수로 반대색 못받는지 검증

addWhite/BlackPawn의 매개변수로 상대편 색을 받았을 때 흰/검의 pawnList에
추가되지 않고 size가 0인 것을 확인하는 테스트를 추가했습니다.

rla36 committed:
try-with resources 추가

Scanner은 AutoClosable을 구현한 클래스이므로 try- with resources를
시도해보았다.

SANGCO reviewed (approved):
고생하셨습니다~
